src/decomposition.jl

```diff
@@ -0,0 +1,167 @@
+"""
+    ear_decomposition(g, s)
+
+Return an Ear decomposition of the graph by using s as source for DFS Tree.
+
+An ear of an undirected unweighted connected graph `G` is a path `P` where
+the two endpoints of the path may coincide (i.e., form a cycle), but where
+otherwise no repetition of edges or vertices is allowed, so every internal
+vertex of `P` has degree two in `P`.
+
+An ear decomposition of an undirected graph `G` is a partition of its
+set of edges into a sequence of ears, such that the one or two endpoints
+of each ear belong to earlier ears in the sequence and such that the
+internal vertices of each ear do not belong to any earlier ear.
+
+For more information, see the 
+[Ear Decomposition](https://en.wikipedia.org/wiki/Ear_decomposition)
+
+It will throw an error if input graph is directed/disconnected graph.
+and return ear_decomposition of connected components if graph is not
+connected.
+
+OUTPUT:
+- A nested array(list) representing the cycles and chains of the ear
+    decomposition of the graph.
+
+### Performance
+Time complexity of this algorithm is `O(N+M)`, where `N` is number of vertices
+and `M` is number of edges.
+
+# Examples
+```jldoctest
+
+julia> using LightGraphs
+
+julia> elist = [(1,2),(2,3),(3,4),(4,1),(1,5), (1, 6), (6, 7), (7, 8), (1, 8)]
+
+julia> for e in elist
+    add_edge!(g, e[1], e[2])
+end
+
+# g is simple undirected connected graph
+julia> ear_decomposition(g, 1)
+2-element Array{Any,1}:
+ [1, 4, 3, 2, 1]
+ [1, 8, 7, 6, 1]
+
+julia> newlist = [(1,2),(2,3),(2,4),(3,4),(4,1), (5, 6), (5, 7), (5, 8), (6, 7), (7, 8)]
+
+julia> g = SimpleGraph(8)
+
+julia>  for e in newlist
+            add_edge!(g, e[1], e[2])
+        end
+
+# g is undirected dis-connected graph
+julia> ear_decomposition(g, 5)
+4-element Array{Array{Int64,1},1}:
+ [5, 7, 6, 5]
+ [5, 8, 7]
+ [1, 4, 3, 2, 1]
+ [2, 4]
+```
+
+### References
+Schmidt, Jens M. (2013a), "A Simple Test on 2-Vertex- and 2-Edge-Connectivity",
+Information Processing Letters, 113 (7): 241–244, arXiv:1209.0700, 
+doi:10.1016/j.ipl.2013.01.016.
+"""
+
+function ear_decomposition end
+@traitfn function ear_decomposition(g::AG::(!IsDirected), s::Integer) where {T, AG<:AbstractGraph{T}}
```

Comment:
I think, we can give `s` the default value 1, as people might often not care, at what vertex we start.